`_find_backend` doesn't work on cached exprs

running
```
import letsql as ls
from letsql.common.caching import ParquetCacheStorage


con = ls.connect()
expr = (
    con.read_parquet(ls.options.pins.get_path("astronauts"))
    .cache(storage=ParquetCacheStorage(source=con))
)
(l, r) = (expr._find_backend().name, con.name)
assert l == r, (l, r)
```
results in 
```
AssertionError                            Traceback (most recent call last)
Cell In[1], line 11
      6 expr = (
      7     con.read_parquet(ls.options.pins.get_path("astronauts"))
      8     .cache(storage=ParquetCacheStorage(source=con))
      9 )
     10 (l, r) = (expr._find_backend().name, con.name)
---> 11 assert l == r, (l, r)

AssertionError: ('duckdb', 'let')
```

--- a/python/letsql/backends/let/hotfix.py
+++ b/python/letsql/backends/let/hotfix.py
@@ -165,7 +165,7 @@ class LETSQLAccessor:
 def _letsql_find_backend(self, *, use_default=True):
     # FIXME: push this into LETSQLAccessor
     try:
-        if tuple(op.source for op in self.op().find(Read)):
+        if tuple(op.source for op in self.op().find((Read, CachedNode))):
             current_backend = _backend_init()
         else:
             current_backend = self._find_backend._original(

--- a/python/letsql/backends/let/hotfix.py
+++ b/python/letsql/backends/let/hotfix.py
@@ -73,6 +73,8 @@ class LETSQLAccessor:
         def _find_backends(expr):
             _backends, _ = expr._find_backends()
             _backends = set(_backends)
+            if backend := expr._find_backend():
+                _backends.add(backend)
 
             for node in expr.op().find_topmost(CachedNode):
                 _backends.update(_find_backends(node.parent))

--- a/python/letsql/backends/let/tests/test_cache.py
+++ b/python/letsql/backends/let/tests/test_cache.py
@@ -940,3 +940,17 @@ def test_ls_exists_doesnt_materialize(cached_two):
     ls.execute(cached_two.count())
     assert cached_two.ls.exists()
     assert storage.exists(cached_two)
+
+
+@pytest.mark.parametrize(
+    "con", [ls.connect(), ls.datafusion.connect(), ls.duckdb.connect()]
+)
+@pytest.mark.parametrize(
+    "cls", [ParquetSnapshot, ParquetCacheStorage, SnapshotStorage, SourceStorage]
+)
+def test_cache_find_backend(con, cls):
+    storage = cls(source=con)
+    expr = con.read_parquet(ls.options.pins.get_path("astronauts")).cache(
+        storage=storage
+    )
+    assert expr._find_backend().name == "let"

--- a/python/letsql/common/collections.py
+++ b/python/letsql/common/collections.py
@@ -4,20 +4,17 @@ from typing import Dict
 
 from ibis.expr.operations import relations as ops
 from letsql.expr.relations import (
+    CachedNode,
     Read,
 )
 
 
 def _find_backend(value):
-    backends = set()
-    node_types = (ops.UnboundTable, ops.DatabaseTable, ops.SQLQueryResult, Read)
-    for table in value.find(node_types):
-        backends.add(table.source)
-
-    if len(backends) > 1:
+    node_types = (ops.UnboundTable, ops.DatabaseTable, ops.SQLQueryResult, CachedNode, Read)
+    (backend, *rest) = set(table.source for table in value.find(node_types))
+    if len(rest) > 1:
         raise ValueError("Multiple backends found for this expression")
-
-    return backends.pop()
+    return backend
 
 
 class SourceDict:

